### PR TITLE
Add Juggernaut to TS

### PR DESCRIPTION
**Pull request by reaperrr**
Collateral changes:
- added `StartBurstReport` support to weapons, because otherwise you'd hear 9(!) shots when the Jugger fires since the soundfile itself is already a triple-shot sound (and yes, this was a bug/design error in original TS/FS)
- added Name field to `WithSpriteBody` to allow assigning modifying traits to specific sprite bodies
- made `WithMoveAnimation` compatible with multiple sprite bodies per actor (to avoid using the `WithInfantryBody` hack for the Juggernaut)
- made `GrantConditionOnDeploy` compatible with multiple sprite bodies per actor and `WithMakeAnimation` work for deploying actors

### CODE REVIEW

File: OpenRA.Mods.Common/Traits/Armament.cs [hunk 1]
```diff
@@ -244,7 +244,9 @@ public virtual Barrel CheckFire(Actor self, IFacing facing, Target target)
 					if (projectile != null)
 						self.World.Add(projectile);
 
-					if (args.Weapon.Report != null && args.Weapon.Report.Any())
+					if (args.Weapon.Report != null && args.Weapon.Report.Any()
+						&& (args.Weapon.OnlyPlayReportOnFirstBurst == false
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Armament.cs]: No need for the explicit comparisons against `true` and `false`.

File: OpenRA.Mods.Common/Traits/Render/WithMoveAnimation.cs [hunk 2]
```diff
@@ -18,6 +19,9 @@ public class WithMoveAnimationInfo : ITraitInfo, Requires<WithSpriteBodyInfo>, R
 		[Desc("Displayed while moving.")]
 		[SequenceReference] public readonly string MoveSequence = "move";
 
+		[Desc("Which sprite body to modify.")]
+		public readonly string AssignedBodyName = "body";
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/WithMoveAnimation.cs]: Could we make this a `readonly string[] BodyNames` to match what we do for armaments?

File: OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnDeploy.cs [hunk 3]
```diff
@@ -57,6 +58,9 @@ public class GrantConditionOnDeployInfo : ITraitInfo
 		[Desc("Can this actor undeploy?")]
 		public readonly bool CanUndeploy = true;
 
+		[Desc("Assigned sprite body (used to play DeployAnimation).")]
+		public readonly string AssignedBodyName = "body";
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnDeploy.cs]: Likewise here.

Comment (pchote) [inline OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnDeploy.cs]: Although ideally we would remove the explicit dependency on `WithSpriteBody` completely, and use a notification instead.

File: mods/ts/weapons/ballisticweapons.yaml [hunk 4]
```diff
@@ -108,6 +108,24 @@ RPGTower:
 		Explosions: large_explosion
 		ImpactSounds: expnew06.aud
 
+Jugg90mm:
```
Comment (pchote) [inline mods/ts/weapons/ballisticweapons.yaml]: Why did you split this out into its own commit?

Comment (reaperrr) [inline mods/ts/weapons/ballisticweapons.yaml]: That's just the order in which I did things, I'll merge it into the main Juggernaut commit.

File: OpenRA.Game/GameRules/WeaponInfo.cs [hunk 5]
```diff
@@ -42,6 +42,8 @@ public sealed class WeaponInfo
 		[Desc("The sound played when the weapon is fired.")]
 		public readonly string[] Report = null;
 
+		public readonly bool OnlyPlayReportOnFirstBurst = false;
```
Comment (GraionDilach) [inline OpenRA.Game/GameRules/WeaponInfo.cs]: Do we really, really need this? This is too much of a hack for me to like. Couldn't the Juggernaut use two weapons with the first weapon being the middle barrel voiced and the two other barrels having a silent Burst: 2 weapon instead?

Comment (reaperrr) [inline OpenRA.Game/GameRules/WeaponInfo.cs]: In my opinion such yaml hacks are not much better, and with two weapons at once you'd have to use FireDelay for the silent weapon on top of that to match the sound's timing with the bursts.
Also, two weapons that actually deal damage aren't great if we ever decide to stray from the "true to original" approach with regard to balancing.

FWIW, I could *really* use this downstream as well.

**Comment by reaperrr** [comment]
Updated. Travis didn't trigger, though.

File: OpenRA.Game/GameRules/WeaponInfo.cs (hunk 5 above)
Comment (pchote) [inline OpenRA.Game/GameRules/WeaponInfo.cs]: IMO you are both right.  This is a bit of a nasty hack, but so is the suggested alternative.

I propose an option c:
```csharp
[Desc("The sound played when the first shot in a burst is fired.")]
public readonly string[] BurstReport = null;
```

File: mods/ts/rules/gdi-vehicles.yaml [hunk 6]
```diff
@@ -279,3 +279,90 @@ SONIC:
...
+		BodyNames: body, deployedbody
+	WithFacingSpriteBody:
+		Sequence: stand
+		RequiresCondition: undeployed
+	WithMoveAnimation:
+		MoveSequence: run
+	GrantConditionOnDeploy:
+		DeployedCondition: deployed
+		UndeployedCondition: undeployed
+		DeployAnimation: make
```
Comment (GraionDilach) [inline mods/ts/rules/gdi-vehicles.yaml]: You don't need this then.

File: mods/ts/rules/nod-vehicles.yaml [hunk 7]
```diff
@@ -190,18 +196,60 @@ ART2:
...
 		LightAmbientColor: 0.4, 0.4, 0.4
-	Transforms:
-		IntoActor: gaarty
+	RenderSprites:
+		Image: art2
+	WithDeployAnimation:
+	GrantConditionOnDeploy:
+		DeployedCondition: deployed
+		UndeployedCondition: undeployed
+		DeployAnimation: make
```
Comment (GraionDilach) [inline mods/ts/rules/nod-vehicles.yaml]: Same.

**Comment by GraionDilach** [comment]
This is a bit problematic in theory. WithDeployAnimation is set up to support multiple WithSpriteBodies, however it will call the finalizers multiple times in that case.

File: OpenRA.Game/GameRules/WeaponInfo.cs (hunk 5 above)
Comment (reaperrr) [inline OpenRA.Game/GameRules/WeaponInfo.cs]: I assume that's meant to be in addition to the existing `Report`?

In my opinion we should at least call it `FirstBurstReport`, as I'm sure not every modder will read the Desc and just `BurstReport` is a bit ambigous (if I was a 3rd-party modder my first assumption would be that it plays for each burst and wonder where's the difference to the normal `Report`.

Also, should this play *in addition to* or *instead of* `Report`?

Comment (GraionDilach) [inline OpenRA.Game/GameRules/WeaponInfo.cs]: In addition.

**Comment by reaperrr** [comment]
> This is a bit problematic in theory. WithDeployAnimation is set up to support multiple WithSpriteBodies, however it will call the finalizers multiple times in that case.

Updated, now only notifies after (un)deploy anim of first enabled WSB has finished.

Also replaced that hacky bool with `FirstBurstReport` as suggested and removed the obsolete yaml lines.

File: OpenRA.Game/GameRules/WeaponInfo.cs [hunk 8]
```diff
@@ -39,9 +39,12 @@ public sealed class WeaponInfo
 		[Desc("The maximum range the weapon can fire.")]
 		public readonly WDist Range = WDist.Zero;
 
-		[Desc("The sound played when the weapon is fired.")]
+		[Desc("The sound played each time the weapon is fired.")]
 		public readonly string[] Report = null;
 
+		[Desc("Sound played only on first shot.")]
+		public readonly string[] FirstBurstReport = null;
```
Comment (pchote) [inline OpenRA.Game/GameRules/WeaponInfo.cs]: Pedantic nit: this suffers from the same ambiguity as `BustReport`, where someone may assume that this sound plays only once ever.  How about `BurstStartReport` or `StartBurstReport`?

Comment (pchote) [inline OpenRA.Game/GameRules/WeaponInfo.cs]: Please also add "in each salvo" or similar to the description, to make it extra clear.

File: OpenRA.Mods.Common/Traits/Render/WithDeployAnimation.cs [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+		[Desc("Sequence name to use.")]
+		[SequenceReference] public readonly string Sequence = "make";
+
+		[Desc("Apply to sprite bodies with these names.")]
+		public readonly string[] BodyNames = { "body" };
+
+		public object Create(ActorInitializer init) { return new WithDeployAnimation(init, this); }
+	}
+
+	public class WithDeployAnimation : INotifyDeployAnimation
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/WithDeployAnimation.cs]: Can we combine this with `WithMakeAnimation`?  (might be too much for this PR).

Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/Render/WithDeployAnimation.cs]: Probably, but I didn't feel like dealing with all the building-specific BuildComplete stuff, considering that (currently) the deploy anim is only needed for actors that don't have/need the Building trait.

I'd also like to at least mention that this merge would be trivial if we got rid of all the BuildComplete hacks first (which is a large undertaking by itself, of course).

**Comment by reaperrr** [comment]
Updated.
Merging the deploy anim into `WithMakeAnimation` turned out to be much easier than I had feared. The only situation in which there may be conflicts is when an actor has the `Building` trait *and* `GrantConditionOnDeploy`, but that's not necessary for any TS actor and will probably be fixed once we get rid of the BuildComplete hacks.
Besides, this also makes WMA multi-`WithSpriteBody`-compatible, which finally allows multiple sprite bodies on buildings.

Additionally, tweaked turret offset to align it with the deploy anim's last frame for seamless transition.

**Comment by pchote** [comment]
The commit comment for "Make WithMoveAnimation multi-WithSpriteBody-compatible" will need to be updated now.

File: OpenRA.Mods.Common/TraitsInterfaces.cs [hunk 10]
```diff
@@ -145,7 +145,24 @@ public interface IProductionIconOverlay
...
+		void AfterTransform(Actor toActor);
+	}
+
+	public interface INotifyDeploy
+	{
+		void FinishedDeploy(Actor self);
+		void FinishedUndeploy(Actor self);
+	}
+
+	public interface INotifyDeployAnimation
```
Comment (pchote) [inline OpenRA.Mods.Common/TraitsInterfaces.cs]: Sorry to keep scope creeping, but `Sellable` and `Transforms` now really should use this instead of hardcoding `WithMakeAnimation`.

Comment (pchote) [inline OpenRA.Mods.Common/TraitsInterfaces.cs]: It feels like this interface really should be called `INotifyDeploy` or perhaps `ITriggerDeploy`, and the old `INotifyDeploy` be something else.

Comment (reaperrr) [inline OpenRA.Mods.Common/TraitsInterfaces.cs]: I have looked at `Sellable` and `Transforms`, and I'm really afraid of the cans of worms that might open. 
I don't mind trying, but please let me do that in its own PR afterwards, because the activity interaction of those two (especially `Sellable`) is just asking for regressions, and this PR is large enough as it is, in my opinion.

Comment (reaperrr) [inline OpenRA.Mods.Common/TraitsInterfaces.cs]: btw, I settled on `INotifyDeployTriggered` for the new and `INotifyDeployComplete` for the 'old' interface.

**Comment by reaperrr** [comment]
Updated and rebased.

File: OpenRA.Mods.Common/TraitsInterfaces.cs (hunk 10 above)
Comment (pchote) [inline OpenRA.Mods.Common/TraitsInterfaces.cs]: Ok, fair enough.

**Comment by reaperrr** [comment]
I'll wait for #13025 to be merged before I rebase this.

**Comment by reaperrr** [comment]
Rebased.

This doesn't yet use the Forward/Reverse for deploy/undeploy in `WithMakeAnimation` that @pchote requested on IRC, if possible I'd like to leave that for a dedicated PR.

File: mods/ts/rules/gdi-vehicles.yaml [hunk 11]
```diff
@@ -280,3 +280,89 @@ SONIC:
...
+	WithSpriteBody@deployed:
+		RequiresCondition: !undeployed && real-actor
+		Name: deployedbody
+	Turreted:
+		Turret: deployed
+		TurnSpeed: 5
+		InitialFacing: 96
+		Offset: -108,-12,448
+		RealignDelay: -1
+	WithVoxelBarrel:
```
Comment (pchote) [inline mods/ts/rules/gdi-vehicles.yaml]: The original game would automatically rotate the pitch of the barrels to match the firing arc.  Doing this properly is very far down my TODO list, so I have filed https://github.com/reaperrr/OpenRA/pull/15 for you to include here that lets hardcode a more sensible fixed angle than horizontal.

**Review by pchote — DISMISSED** [review]
We're converging to a good solution here, but I have a couple of final requests:

**Comment by reaperrr** [comment]
Updated and rebased, also applied the barrel angle fixes to Nod artillery.

File: OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnDeploy.cs [hunk 12]
```diff
@@ -42,9 +43,6 @@ public class GrantConditionOnDeployInfo : ITraitInfo
 		[Desc("Cursor to display when unable to (un)deploy the actor.")]
 		public readonly string DeployBlockedCursor = "deploy-blocked";
 
-		[SequenceReference, Desc("Animation to play for deploying/undeploying.")]
-		public readonly string DeployAnimation = null;
```
Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnDeploy.cs]: Done.

File: mods/ts/rules/gdi-vehicles.yaml [hunk 13]
```diff
@@ -280,3 +280,89 @@ SONIC:
 		Offset: -170,0,0
 	AutoTarget:
 	WithVoxelTurret:
+
+JUGG:
```
Comment (reaperrr) [inline mods/ts/rules/gdi-vehicles.yaml]: sel-box

Comment (reaperrr) [inline mods/ts/rules/gdi-vehicles.yaml]: This is how it looks now.

**Review by pchote — APPROVED** [review]

**Comment by pchote** [comment]
I noticed one final problem while testing this, but I think at this point it is better left for a future PR: the turret needs to rotate back to the south west before triggering the undeploy animation.

**Comment by pchote** [comment]
Now also needs a rebase.

**Comment by reaperrr** [comment]
Depends on #13253, so I can adapt the LocalOffsets in this PR.

**Comment by atlimit8** [comment]
#13253 has been merged.

**Comment by reaperrr** [comment]
Rebased, and adapted the last two commits to isometric WDist rescaling.

**Comment by pchote** [comment]
Ping @OpenRA/engine-hackers.  This PR is holding up a replacement make animation / buildup trait that @IceReaper is working on for the kknd and mw mods.

**Comment by reaperrr** [comment]
Rebased, but I suspect the sheer size of the PR is scaring off potential reviewers so I'll split off parts into more digestable PRs to hopefully get the crucial bits merged more quickly.

**Comment by reaperrr** [comment]
Rebased. This PR now consists of a single commit that does nothing more than what the title suggests.

**Review by IceReaper — APPROVED** [review]
